aws-sdk-core/lib/aws-sdk-core/shared_credentials.rb

```diff
@@ -67,15 +67,35 @@ def default_path
 
     def load_from_path
       profile = load_profile
-      @credentials = Credentials.new(
+      credentials = Credentials.new(
         profile['aws_access_key_id'],
         profile['aws_secret_access_key'],
         profile['aws_session_token']
       )
+      @credentials = if role_arn = profile['role_arn']
+        AssumeRoleCredentials.new(
+          role_session_name: profile['role_session_name'] || "AWS-Ruby-session-%s"%[Time.now.to_i],
+          external_id: profile['external_id'],
+          role_arn: role_arn,
+          credentials: credentials
+        ).credentials
+      else
+        credentials
+      end
     end
 
     def load_profile
       if profile = profiles[profile_name]
+        if profile.key?('role_arn')
+          if source_profile = profile['source_profile']
+            profile['aws_access_key_id'] = profiles[source_profile]['aws_access_key_id']
```

Comment:
But what happens if we will have the defined one from these parameters in non-source profile? Seems they will be re-written from source one. I believe the usage of method "merge" is more correct. But I agree with check the existence of  source profile.